tests/end2end-tests/graphbolt-gs-integration/graphbolt-training-inference.sh

```diff
@@ -0,0 +1,249 @@
+# Ensure ip_list.txt exists and self-ssh works
+rm "$OUTPUT_PATH/ip_list.txt" &> /dev/null || true
+echo "127.0.0.1" > "$OUTPUT_PATH/ip_list.txt"
+ssh -o PreferredAuthentications=publickey -o StrictHostKeyChecking=no \
+    -p 2222 127.0.0.1 /bin/true || service ssh restart
+
+# Generate 1P LP data
+LP_INPUT_1P="${OUTPUT_PATH}/graphbolt-gconstruct-lp-1p"
+python3 -m graphstorm.gconstruct.construct_graph \
+    --add-reverse-edges \
+    --conf-file $GS_HOME/tests/end2end-tests/data_gen/movielens_lp.json \
+    --graph-name ml-lp \
+    --num-parts 1 \
+    --num-processes 1 \
+    --output-dir "$LP_INPUT_1P" \
+    --part-method random \
+    --use-graphbolt "true"
+
+LP_OUTPUT="$OUTPUT_PATH/gb-lp"
+msg "**************GraphBolt Link Prediction training. dataset: Movielens, RGCN layer 1, node feat: fixed HF BERT, BERT nodes: movie, inference: mini-batch, negative_sampler: joint, exclude_training_targets: false"
+python3 -m graphstorm.run.gs_link_prediction \
+    --cf $GS_HOME/training_scripts/gsgnn_lp/ml_lp.yaml \
+    --eval-frequency 300 \
+    --ip-config "$OUTPUT_PATH/ip_list.txt" \
+    --num-epochs 1 \
+    --num-samplers 0 \
+    --num-servers 1 \
+    --num-trainers 1 \
+    --part-config "$LP_INPUT_1P/ml-lp.json" \
+    --save-model-path "$LP_OUTPUT/model" \
+    --ssh-port 2222 \
+    --use-graphbolt true
+
+# Ensure model file was saved
+fdir_exists f "$LP_OUTPUT/model/epoch-0/model.bin"
+
+msg " **************GraphBolt Link Prediction embedding generation **************"
+
+python3 -m graphstorm.run.gs_gen_node_embedding \
+    --cf $GS_HOME/training_scripts/gsgnn_lp/ml_lp.yaml \
+    --eval-frequency 300 \
+    --inference \
+    --ip-config "$OUTPUT_PATH/ip_list.txt" \
+    --num-epochs 1 \
+    --num-samplers 0 \
+    --num-servers 1 \
+    --num-trainers 1 \
+    --part-config "$LP_INPUT_1P/ml-lp.json" \
+    --restore-model-path "$LP_OUTPUT/model/epoch-0" \
+    --save-embed-path "$LP_OUTPUT/embeddings" \
+    --ssh-port 2222 \
+    --use-graphbolt true
+
+# Ensure embeddings were created
+fdir_exists d "$LP_OUTPUT/embeddings/movie"
+fdir_exists d "$LP_OUTPUT/embeddings/user"
+
+LP_OUTPUT="$OUTPUT_PATH/gb-lp-inbatch_joint"
+msg "**************GraphBolt Link Prediction training. dataset: Movielens, RGCN layer 1, inference: mini-batch, negative_sampler: inbatch_joint, exclude_training_targets: true"
+python3 -m graphstorm.run.gs_link_prediction \
+    --cf $GS_HOME/training_scripts/gsgnn_lp/ml_lp.yaml \
+    --eval-frequency 300 \
+    --exclude-training-targets True \
+    --ip-config "$OUTPUT_PATH/ip_list.txt" \
+    --num-epochs 1 \
+    --num-samplers 0 \
+    --num-servers 1 \
+    --num-trainers 1 \
+    --part-config "$LP_INPUT_1P/ml-lp.json" \
+    --reverse-edge-types-map user,rating,rating-rev,movie \
+    --save-model-path "$LP_OUTPUT/model" \
+    --ssh-port 2222 \
+    --train-negative-sampler inbatch_joint \
+    --use-graphbolt true
+
+# Ensure model file was saved
+fdir_exists f "$LP_OUTPUT/model/epoch-0/model.bin"
```

Comment:
Do we store the optimizer here?

Reply:
I believe so, I could check for that too.